### PR TITLE
Write down reasoning behind conditionals decision

### PR DESCRIPTION
**Pull request by kaushik-work**
Current prototype implementation: https://github.com/DataBiosphere/toil/pull/2845

@mr-c I felt it better to have this in the specifications repository instead of floating around in yet another repository. Thanks!

(background: https://docs.google.com/document/d/1-6kg8r6bUG-ENmZoCqvKFGpfNL0kWPX7YT63Awb3kOE/edit?ts=5cf68120

specifically
> Should we document this process using Architecture Decision Records? https://github.com/joelparkerhenderson/architecture_decision_record

)

### CODE REVIEW

File: design-decisions/conditionals-2019.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+## Major decision points (optimizations)
+
+- Minimize disruption to current syntax
+- Minimize the number of new concepts exposed to the user
+- Minimize amount of text developer has to write
+- Make auditable (make it easily apparent what is a conditional and in what way)
+- Minimize the number of new concepts exposed to the runner developer
+
+The new syntax adds a single field to `WorkflowStep` and a new operator called `branchSelect`
```
[the remark this answers is not in the document]
Comment (stain) [reply design-decisions/conditionals-2019.md]: The proposal in #854 suggests using a variant of link merging as part of conditionals. The reason for this should be better reflected in the rationale here (beyond just "not a nested workflow").

File: design-decisions/conditionals-2019.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+## Need for conditionals in CWL
+
+The initiative to implement conditionals in CWL has a long history. 
+There is a long and illuminating thread on the [CWL mailing list](https://groups.google.com/d/topic/common-workflow-language/JU7PSEKr97M/discussion)
+on this topic. There have been several proposals on the [github issues section](https://github.com/common-workflow-language/common-workflow-language/issues?q=is%3Aissue+is%3Aopen+label%3Aconditionals)
+
+In short, while there is a minority sentiment that CWL should remain a pure dataflow 
+language with no conditionals, there is a majority sentiment that, in the face of
+reality, users need the ability to instruct the CWL runner on conditional execution.
+
```
[the remark this answers is not in the document]
Comment (kaushik-work) [reply design-decisions/conditionals-2019.md]: @stain I'm not sure about "predictable dataflow semantics". I think I know what it means but could you give an example that is not predictable. i.e. what can we do to the DAG that makes it unpredictable? Even adding loops makes the data flow predictable in my view, for example.

Does this have the same meaning as "explicit dataflow"? i.e. you can look at the DAG and infer what it does?

In such a case scatter for me is pretty non-intuitive. It took me a while before I learned to mentally explode a node with scatter into a set of parallel nodes with indeterminate number. And I still have difficulty with chained nodes that each scatter. 

Thanks.

File: design-decisions/conditionals-2019.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (kaushik-work) [reply design-decisions/conditionals-2019.md]: Yes! This is the most important thing, even! I'll put it in.

**Comment by jmchilton** [comment]
Thanks for laying this all out - I think the document is super useful and I appreciate all the effort. I think what is here makes some good conceptual sense outside the context of scatter/merge. I'm nervous about thinking through scatter/merge - but that is probably a deficiency with my ability to keep everything in my head. That said, if this included an explicit discussion of scatter and merging and how things work out with examples that would be really helpful.

**Comment by kaushik-work** [comment]
@jmchilton great idea to include how scatter and merge are handled. This is important, and I believe it works rationally - so will be good to get feedback on this. Thanks!

**Comment by jmchilton** [comment]
I've slept on this and I've decided this should be simpler and more general.

> all outputs of a skipped step that are otherwise not handled are converted to the CWL `null` type. 

This feels right - I like this! Given this I don't think we really should distinguish between a skipped output defaulted to null and an explicit null output. My argument here is essentially "who cares?" - my default position is this is complexity neither the user nor the implementer really *needs*. If we don't need it, lets try without it. 

One of the design goals with this was to avoid thinking about state - we aren’t exposing a ‘skipped’ states in CWL documents or to CWL authors at all. This again was good, and I think my proposal here to eliminate the concept of branch selection and replace it with just dispatching on nulls goes farther in that direction. 

We are eliminating ideas we expose to the user I think - they don’t really need to think about whether a previous step was skipped or just null. This reduces the cognitive load. And if 99.98% of the time CWL authors won’t care about the distinction between an explicit null and a skipped implicit null (my opinion), we should eliminate the cognitive load and implementation complexity associated with it.

In addition to simplifying this for implementers and users, I think the result is constructs that are more general and reusable. This is good. For instance:

```
source:
  first_that_ran:
    - step1/out1
    - in1
```

becomes:

```
source:
  first_non_null:
    - step1/out1
    - in1
```

I think this name is more clear and that feels to me like evidence the cognitive load is smaller. This also becomes a more general purpose tool that has uses outside of conditionals. We could implement optional exclusive parameters more cleanly using this, dispatching on filtered arrays, etc.. We’re finally taking steps down the road toward replacing common expressions with small utility patterns. 

Another benefit of this new approach here is ``first_non_null`` and ``when`` can now be implemented and added independently to the spec. They aren’t needed for each other and are useful independent of each other.

**Comment by kaushik-work** [comment]
@jmchilton I like the idea of stripping away as much as possible. `first_non_null` and `remove_nulls` would do the job, and do it in an easy to understand manner. The crucial point here is what use cases we are impeding by doing this. You have convinced me that `null` outputs are likely a tiny minor (mis-)use case.

**Comment by stain** [comment]
I also like this late edition from @jmchilton - it is very important to state that it is `first_non_null` rather than `first_that_ran`  as that would implies a temporal order. In fact this is exactly how we used to do this in Taverna 1 where we picked the value that first (in time) appeared on the input - which could give unpredictable outputs in the case where multiple upstreams ran.

We found it was was easy for users to get multiple conditionals wrong - e.g. they would have two steps, one with "if a <1" and aother "if a>1" and not cover "a==1".  Or what we saw more, they would do "if not a>1" and "if not a<1"  and both get triggered on "a==1" in unpredictable order.  

Add multiple inputs to evaluate and it's not easy to see the conditionals that can trigger at same time.

So the name is important - I think we want to have **in pre-defined order** as an important requirement here.

Now back to the proposal from @jmchilton :

If someone really wanted the ability to return null that would not trigger fallover, they can change their types to `list` and return `[null]` or `null`  - basically the existing array optional hack.

Actually I think this is an advantage, as it means the step itself can decide to fail, even after its execution, e.g. a `valueFrom` can decide the value was not good enough and refuse to deliver its output (give `null`) instead, triggering the next non-null value to be preferred.

It does mean unnecessary execution of the additional upstream fallback steps (we have not in the spec said that an engine is free to **not** execute steps that do not contribute to `outputs` - they may be doing house keeping).

**Comment by kaushik-work** [comment]
@jmchilton , @stain, @tetron @mr-c  I've been thinking about the syntax:

```
source:
  first_non_null:
    - step1/out1
    - in1
```

And now it is stuck in my head that this operator is really of the same kind as the `link_merge` operator. The way the syntax is denoted here is nice, but a little arbitrary. Why is `link_merge` sitting outside? Also, this means source can now be a string or an object, which is not the worst thing, but a little irregular.

I'd like to go back to the earlier syntax, which was congruent with how we had `link_merge`. So:

```
source:
    - step1/out1
    - in1
null_filter: first_non_null
link_merge: merge_flattened
```

A future extension can have `fail_filter` and so on.
I feel what we are doing here is applying different operators to this list of inputs and producing another list or a scalar.

In terms of operator precedence, we have well defined rules and they are (including this proposal)

`null_filter -> link_merge -> default -> valueFrom`

**Comment by kaushik-work** [comment]
This has been moved to the CWL 1.2 repository -> https://github.com/common-workflow-language/cwl-v1.2/pull/4

**Pull request closed without merging**